tmt/steps/report/reportportal.py

```diff
@@ -38,113 +25,179 @@ class ReportReportPortalData(tmt.steps.report.ReportStepData):
         option="--token",
         metavar="TOKEN",
         default=os.environ.get("TMT_REPORT_REPORTPORTAL_TOKEN"),
-        help="The token to use for upload to the ReportPortal instance.")
-    project: Optional[str] = tmt.utils.field(
+        help="The token to use for upload to the ReportPortal instance (from RP user profile).")
+    project: Optional[str] = tmt.utils.field(  # todo: argument should be mandatory
         option="--project",
-        metavar="PROJECT",
+        metavar="PROJECT_NAME",
         default=None,
         help="The project name which is used to create the full URL.")
-    launch_name: Optional[str] = tmt.utils.field(
-        option="--launch-name",
-        metavar="NAME",
+    launch: Optional[str] = tmt.utils.field(
+        option="--launch",
+        metavar="LAUNCH_NAME",
+        default=None,
+        help="The launch name (name of plan per launch is used by default).")
+    env_var: Optional[str] = tmt.utils.field(
+        option="--env-var",
+        metavar="TMT_VAR",
         default=None,
-        help="The launch name (base name of run id used by default).")
+        multiple=True,
```

Comment:
@happz This is where I struggle. I've tried several options and none works as I'd expected. This is the behaviour with changes you suggested:
* without 'default'
```
    env_var: List[str] = tmt.utils.field(
        option="--env-var",
        metavar="TMT_VAR",
        normalize=tmt.utils.normalize_string_list,
        multiple=True,
        help="...")
```
```
$ tmt
Container field must define one of 'default' or 'default_factory'.
```
Seems the field() needs the default there, although I would rather not have it here in 'env-var' arg as well as in 'project' arg so I can omit the optionality.

* with 'default' and the type `List[str]`
```
env_var: List[str] = tmt.utils.field(
        option="--env-var",
        metavar="TMT_VAR",
        normalize=tmt.utils.normalize_string_list,
        default=None,
        multiple=True,
        help="...")
```

1.  with no string in '.fmf' or CLI for env-var
```
$ tmt run report
...
report
        how: reportportal
            tmt_env_vars: []
...
```

2.1  with one string in '.fmf' for env-var
```
$ tmt run report
...
    report
        how: reportportal
            tmt_env_vars: ['TMT_TREE']
...
```

2.2  with two strings in '.fmf' for env-var
```
$ tmt run report
...
report
        how: reportportal
            tmt_env_vars: ['TMT_PLAN_DATA TMT_TREE']
...
```
3.1 with one string in CLI for env-var
```
 $ tmt run report --how reportportal --env-var TMT_TREE 
...
report
        how: reportportal
            tmt_env_vars: ('TMT_TREE',)
...
```

3.2 with two strings in CLI for env-var
```
$ tmt run report --how reportportal --env-var TMT_TREE TMT_PLAN_DATA
Usage: tmt run [OPTIONS] COMMAND1 [ARGS]... [COMMAND2 [ARGS]...]...
Try 'tmt run --help' for help.

Error: No such command 'TMT_PLAN_DATA'.
```
Seems like the multiple option doesn't work at all. And only 3.1 is in the format that works (RP shows what I expect).

* with the original type `Optional[str]` :
```
    env_var: Optional[str] = tmt.utils.field(
        option="--env-var",
        metavar="TMT_VAR",
        default=None,
        multiple=True,
        help="...")
```

4.1 with one string in '.fmf' for env-var
```
$ tmt run report
...
report
        how: reportportal
            tmt_env_vars: TMT_TREE
...
```

4.2 with two strings in '.fmf' for env-var
```
$ tmt run report
...
    report
        how: reportportal
            tmt_env_vars: TMT_PLAN_DATA TMT_TREE
...
```

5.1 with one string in CLI for env-var 
--> same as 3.1

5.2 with two strings in CLI for env-var 
--> same as 3.2

Though I get unstable format, at least it does work (RP shows what I expect), except the 5.2.

Reply:
> @happz This is where I struggle. I've tried several options and none works as I'd expected. This is the behaviour with changes you suggested:
> 
> * without 'default'
> 
> ```
>     env_var: List[str] = tmt.utils.field(
>         option="--env-var",
>         metavar="TMT_VAR",
>         normalize=tmt.utils.normalize_string_list,
>         multiple=True,
>         help="...")
> ```

`default_factory=list` should help here. There cannot be a default value, but there can be a factory callable that creates such a default value when field is initialized.

> 
> ```
> $ tmt
> Container field must define one of 'default' or 'default_factory'.
> ```

Correct behavior, a default of some kind is needed.

> 
> Seems the field() needs the default there, although I would rather not have it here in 'env-var' arg as well as in 'project' arg so I can omit the optionality.

The "optionality" is not needed with `List[str]` - the value would *always* be a list. If user did not specify any value, it would be an empty list, signaling "not set", "no env vars to deal with". 

> 
> * with 'default' and the type `List[str]`
> 
> ```
> env_var: List[str] = tmt.utils.field(
>         option="--env-var",
>         metavar="TMT_VAR",
>         normalize=tmt.utils.normalize_string_list,
>         default=None,
>         multiple=True,
>         help="...")
> ```

`default=None` is in conflict with `List[str]` here: the annotation does not allow the value to be `Optional`, and the value shall be a list, both of these rules would be violated by `None`.

> 
> 1. with no string in '.fmf' or CLI for env-var
> 
> ```
> $ tmt run report
> ...
> report
>         how: reportportal
>             tmt_env_vars: []
> ...
> ```
> 
> 2.1 with one string in '.fmf' for env-var
> 
> ```
> $ tmt run report
> ...
>     report
>         how: reportportal
>             tmt_env_vars: ['TMT_TREE']
> ...
> ```
> 
> 2.2 with two strings in '.fmf' for env-var
> 
> ```
> $ tmt run report
> ...
> report
>         how: reportportal
>             tmt_env_vars: ['TMT_PLAN_DATA TMT_TREE']
> ...
> ```
> 
> 3.1 with one string in CLI for env-var
> 
> ```
>  $ tmt run report --how reportportal --env-var TMT_TREE 
> ...
> report
>         how: reportportal
>             tmt_env_vars: ('TMT_TREE',)
> ...
> ```
> 
> 3.2 with two strings in CLI for env-var
> 
> ```
> $ tmt run report --how reportportal --env-var TMT_TREE TMT_PLAN_DATA
> Usage: tmt run [OPTIONS] COMMAND1 [ARGS]... [COMMAND2 [ARGS]...]...
> Try 'tmt run --help' for help.
> 
> Error: No such command 'TMT_PLAN_DATA'.
> ```
> 
> Seems like the multiple option doesn't work at all. And only 3.1 is in the format that works (RP shows what I expect).
> 
> * with the original type `Optional[str]` :
> 
> ```
>     env_var: Optional[str] = tmt.utils.field(
>         option="--env-var",
>         metavar="TMT_VAR",
>         default=None,
>         multiple=True,
>         help="...")
> ```
> 
> 4.1 with one string in '.fmf' for env-var
> 
> ```
> $ tmt run report
> ...
> report
>         how: reportportal
>             tmt_env_vars: TMT_TREE
> ...
> ```
> 
> 4.2 with two strings in '.fmf' for env-var
> 
> ```
> $ tmt run report
> ...
>     report
>         how: reportportal
>             tmt_env_vars: TMT_PLAN_DATA TMT_TREE
> ...
> ```
> 
> 5.1 with one string in CLI for env-var --> same as 3.1
> 
> 5.2 with two strings in CLI for env-var --> same as 3.2
> 
> Though I get unstable format, at least it does work (RP shows what I expect), except the 5.2.

Please, try you experiments once again, with `default_factory=list` and `List[str]`. The field is supposed to be "a list of envvar names" - tmt is perfectly capable to handle it as a list. What it cannot do is split names in a string, e.g. `--env-var "TMT_FOO TMT_BAR"` would end up as `env_var == ['TMT_FOO TMT_BAR']`.

Reply:
I worry, it is still the same..
How I understand it, `--env-var TMT_VAR_1 TMT_VAR_2 TMT_VAR_3` in CLI and `env-var: TMT_VAR_1 TMT_VAR_2 TMT_VAR_3` in '.fmf' is supposed to be perfectly valid example for tmt list.
But CLI still does not accept more than one argument after --env-var and all env-var variables from .fmf end up as one string in the list.

Reply:
@happz Thank you for that! You're right..
I hoped, indentation works the same in fmf data works the same at any layer, and overlooked the need of the dashes there..
I'm sorry for the misunderstanding.